README.md
aldryn-background-image
=======================

Plugin that allows to add background-images.


CHANGELOG.rst
Changelog
=========

0.1.0 (2016-02-09)
------------------

* Initial release


aldryn_background_image/cms_plugins.py
# -*- coding: utf-8 -*-

from __future__ import unicode_literals

from django.utils.translation import ugettext_lazy as _

from cms.plugin_pool import plugin_pool
from cms.plugin_base import CMSPluginBase

from .models import BackgroundImage


class BackgroundImagePlugin(CMSPluginBase):
    model = BackgroundImage
    name = _("Background Image")
    render_template = "aldryn_background_image/plugin.html"
    allow_children = True

    fieldsets = (
        (None, {
            'fields': ('background_image', 'tag_type', )
        }),
        (_('Advanced Settings'), {
            'classes': ('collapse',),
            'fields': (
                'background_repeat',
                'background_position',
                'background_size',
                'additional_class_names',
                'label',
                'id_name',
            ),
        }),
    )

plugin_pool.register_plugin(BackgroundImagePlugin)


aldryn_background_image/migrations/0001_initial.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import migrations, models
import filer.fields.image
import django.db.models.deletion


class Migration(migrations.Migration):

    dependencies = [
        ('cms', '0013_urlconfrevision'),
        ('filer', '0002_auto_20150606_2003'),
    ]

    operations = [
        migrations.CreateModel(
            name='BackgroundImage',
            fields=[
                ('cmsplugin_ptr', models.OneToOneField(parent_link=True, related_name='+', primary_key=True, serialize=False, to='cms.CMSPlugin')),
                ('label', models.CharField(help_text='Optional label for this plugin.', max_length=128, verbose_name='label', blank=True)),
                ('id_name', models.CharField(max_length=50, verbose_name='id name', blank=True)),
                ('tag_type', models.CharField(default='div', max_length=50, verbose_name='tag Type', choices=[('div', 'div'), ('header', 'header'), ('footer', 'footer'), ('article', 'article'), ('section', 'section')])),
                ('background_repeat', models.CharField(default='repeat', max_length=128, null=True, blank=True, choices=[('repeat', 'repeat'), ('repeat-x', 'repeat-x'), ('repeat-y', 'repeat-y'), ('no-repeat', 'no-repeat'), ('initial', 'initial'), ('inherit', 'inherit')])),
                ('background_position', models.CharField(max_length=128, null=True, blank=True)),
                ('background_size', models.CharField(help_text='auto|length|cover|contain|initial|inherit', max_length=128, null=True, blank=True)),
                ('additional_class_names', models.TextField(help_text='Comma separated list of additional classes to apply to tag_type', verbose_name='additional classes', blank=True)),
                ('background_image', filer.fields.image.FilerImageField(on_delete=django.db.models.deletion.SET_NULL, to='filer.Image', null=True)),
            ],
            options={
                'abstract': False,
            },
            bases=('cms.cmsplugin',),
        ),
    ]


aldryn_background_image/models.py
# -*- coding: utf-8 -*-

from __future__ import unicode_literals

import re

from django.core.exceptions import ValidationError
from django.db import models
from django.utils.encoding import python_2_unicode_compatible
from django.utils.translation import ugettext_lazy as _

from cms.models import CMSPlugin
from filer.fields.image import FilerImageField


CLASS_NAME_FORMAT = re.compile(r'^\w[\w_-]*$')

@python_2_unicode_compatible
class BackgroundImage(CMSPlugin):
    """
    A django CMS Plugin to change background images
    """

    TAG_TYPE_CHOICES = (
        ('div', 'div'),
        ('header', 'header'),
        ('footer', 'footer'),
        ('article', 'article'),
        ('section', 'section'),
    )
    BACKGROUND_REPEAT_CHOICES = (
        ('repeat', 'repeat'),
        ('repeat-x', 'repeat-x'),
        ('repeat-y', 'repeat-y'),
        ('no-repeat', 'no-repeat'),
        ('initial', 'initial'),
        ('inherit', 'inherit'),
    )

    cmsplugin_ptr = models.OneToOneField(
        CMSPlugin, related_name='+', parent_link=True)

    label = models.CharField(_('label'), max_length=128, blank=True,
                             help_text=_('Optional label for this plugin.'))
    id_name = models.CharField(_('id name'), max_length=50, blank=True)
    tag_type = models.CharField(
        verbose_name=_('tag Type'), max_length=50, choices=TAG_TYPE_CHOICES,
        default=TAG_TYPE_CHOICES[0][0])
    background_image = FilerImageField(null=True, on_delete=models.SET_NULL)
    background_repeat = models.CharField(
            max_length=128, blank=True, null=True, default=BACKGROUND_REPEAT_CHOICES[0][0],
            choices=BACKGROUND_REPEAT_CHOICES)
    background_position = models.CharField(max_length=128, blank=True,
                                           null=True)
    background_size = models.CharField(
            max_length=128, blank=True, null=True,
            help_text=_('auto|length|cover|contain|initial|inherit'))

    additional_class_names = models.TextField(
        verbose_name=_('additional classes'),
        blank=True,
        help_text=_('Comma separated list of additional classes to apply to '
                    'tag_type')
    )

    def __str__(self):
        display = self.tag_type or ''
        if self.additional_class_names:
            display = '{0} ({1})'.format(display, self.additional_class_names)
        if self.label:
            display = '“{0}”: {1}'.format(self.label, display)
        return display

    def clean(self):
        if self.additional_class_names:
            additional_class_names = list(
                html_class.strip() for html_class in
                self.additional_class_names.split(','))
            for class_name in additional_class_names:
                class_name = class_name.strip()
                if not CLASS_NAME_FORMAT.match(class_name):
                    raise ValidationError(
                        _('"{0}" is not a proper css class name.').format(
                            class_name)
                    )
            self.additional_class_names = ', '.join(
                set(additional_class_names))

    @property
    def get_additional_class_names(self):
        if self.additional_class_names:
            # Removes any extra spaces
            return ' '.join((
                html_class.strip() for html_class in
                self.additional_class_names.split(',')))
        return ''
